mix/CodeModel.cpp

```diff
@@ -327,6 +331,46 @@ void CodeModel::runCompilationJob(int _jobId)
 	emit stateChanged();
 }
 
+void CodeModel::gasEstimation(solidity::CompilerStack const& _cs)
+{
+	m_gasCostsMaps.clear();
+	for (std::string n: _cs.getContractNames())
+	{
+		ContractDefinition const& contractDefinition = _cs.getContractDefinition(n);
+		QString sourceName = QString::fromStdString(*contractDefinition.getLocation().sourceName);
+
+		if (!m_gasCostsMaps.contains(sourceName))
+			m_gasCostsMaps.insert(sourceName, QVariantList());
+
+		if (!contractDefinition.isFullyImplemented())
+			continue;
+		dev::solidity::SourceUnit const& sourceUnit = _cs.getAST(*contractDefinition.getLocation().sourceName);
+		AssemblyItems const* items = _cs.getRuntimeAssemblyItems(n);
+
+		StructuralGasEstimator estimator;
+		std::map<ASTNode const*, GasMeter::GasConsumption> gasCosts = estimator.breakToStatementLevel(estimator.performEstimation(*items, std::vector<ASTNode const*>({&sourceUnit})), {&sourceUnit});
+
+		for (auto gasItem = gasCosts.begin(); gasItem != gasCosts.end(); ++gasItem)
+		{
+			SourceLocation const& location = gasItem->first->getLocation();
+			GasMeter::GasConsumption cost = gasItem->second;
+			std::stringstream v;
+			v << cost.value;
+			GasMap* gas = new GasMap(location.start, location.end, QString::fromStdString(v.str()), cost.isInfinite);
```

Comment:
In the ctor, the ownership is set to JavaScript. So I think it's enough no?

2015-05-15 9:23 GMT-07:00 Arkadiy Paronyan <notifications@github.com>:

> In mix/CodeModel.cpp
> <https://github.com/ethereum/cpp-ethereum/pull/1923#discussion_r30423015>:
>
> > +
> > +		if (!contractDefinition.isFullyImplemented())
> > +			continue;
> > +		dev::solidity::SourceUnit const& sourceUnit = _cs.getAST(*contractDefinition.getLocation().sourceName);
> > +		AssemblyItems const* items = _cs.getRuntimeAssemblyItems(n);
> > +
> > +		StructuralGasEstimator estimator;
> > +		std::map<ASTNode const*, GasMeter::GasConsumption> gasCosts = estimator.breakToStatementLevel(estimator.performEstimation(*items, std::vector<ASTNode const*>({&sourceUnit})), {&sourceUnit});
> > +
> > +		for (auto gasItem = gasCosts.begin(); gasItem != gasCosts.end(); ++gasItem)
> > +		{
> > +			SourceLocation const& location = gasItem->first->getLocation();
> > +			GasMeter::GasConsumption cost = gasItem->second;
> > +			std::stringstream v;
> > +			v << cost.value;
> > +			GasMap* gas = new GasMap(location.start, location.end, QString::fromStdString(v.str()), cost.isInfinite);
>
> This still needs to be deleted later, otherwize it is a memory leak.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ethereum/cpp-ethereum/pull/1923/files#r30423015>.
>